// phosphor-dgrid/src/BaseGrid.ts
// This file is part of phosphor-dgrid, copyright (c) 2017-2018 BusFaster Ltd.
// Released under the MIT license, see LICENSE.

import 'dgrid/css/dgrid.css';

import 'dojo';
import * as declare from 'dojo/_base/declare';
import * as OnDemandGrid from 'dgrid/OnDemandGrid';
import * as Keyboard from 'dgrid/Keyboard';
import * as Selection from 'dgrid/Selection';
import * as ColumnResizer from 'dgrid/extensions/ColumnResizer';

import { Widget } from '@phosphor/widgets';

export type BaseGridOptions = (
	OnDemandGrid.KwArgs &
	Keyboard.KwArgs &
	Selection.KwArgs &
	ColumnResizer.KwArgs
);

export class BaseGrid extends Widget {

	constructor(
		options: BaseGridOptions,
		GridType = BaseGrid.GridClass
	) {
		super({ node: BaseGrid.createNode() });

		this.addClass('phosphor-dgrid');
		this.addClass('charto-panel');

		const dgrid = new GridType(options, document.createElement('div'));
		this.dgrid = dgrid;

		dgrid.domNode.classList.add('charto-content');
		// Override height set by dgrid.
		dgrid.domNode.style.height = '100%';

		this.node.appendChild(dgrid.domNode);
		dgrid.startup();
	}

	static createNode(): HTMLElement {
		return(document.createElement('div'));
	}

	onResize(msg: Widget.ResizeMessage) {
		if(this.dgrid) this.dgrid.resize();
	}

	static GridClass = declare([ OnDemandGrid, Keyboard, Selection, ColumnResizer ]);

	dgrid: OnDemandGrid & Keyboard & Selection & ColumnResizer;

}
